### ISSUE TITLE
Restricted users loading all operations causing 500

### ISSUE THREAD

**Issue by edlerd** [canonical/lxd#13576]
# Required information

 * Distribution: snap
 * Distribution version: 5.21

# Issue description

On a `GET 1.0/operations?all-projects=true&recursion=1` from a user with tls authentication and `restricted: true` I get a 500 response with this message: `Failed to get operation permission checker: Certificate is restricted error_code: 500`. I would expect a list of operations from all projects, the user has access to.

Removing the `all-projects=true` flag results in a 200 response with correct data.

This seems related to #13533

# Steps to reproduce

 1. Create a custom project
 2. Create a TLS user for the UI
 3. Restrict the TLS user to the custom project
 4. Request `GET /1.0/operations?all-projects=true&recursion=1`

**Comment by tomponline** [comment]
@markylaing please can you take a look at this too.

**Comment by markylaing** [comment]
@edlerd is this a regression?

As far as I know, it has never been possible for a restricted client to list resources of any kind with `all-projects=true`.

Example:
```
$ lxc config trust show "tls:$(lxc query tls:/1.0 | jq -r '.auth_user_name')"
name: mark
type: client
restricted: true
projects:
- default
certificate: |
  -----BEGIN CERTIFICATE-----
  MIIBtjCCATygAwIBAgIQBIHH574cQbVm2RF5CLbl3jAKBggqhkjOPQQDAzAjMQww
  CgYDVQQKEwNMWEQxEzARBgNVBAMMCm1hcmtAUlVCSVgwHhcNMjQxMDE0MTcwMzEy
  WhcNMzQxMDEyMTcwMzEyWjAjMQwwCgYDVQQKEwNMWEQxEzARBgNVBAMMCm1hcmtA
  UlVCSVgwdjAQBgcqhkjOPQIBBgUrgQQAIgNiAATa+hyuRheuVhhEki/Tm6XLMZBS
  XF3juZvIjwfUAktIBGnjUoFw+Rxa3Q0H048u23W+AOABXm1x88Vl0ZrgNQFZ6I6k
  SEoN5RaFPp0QTmJmX9mfLQtjfgbZ1PKo+5ZiKHCjNTAzMA4GA1UdDwEB/wQEAwIF
  oDATBgNVHSUEDDAKBggrBgEFBQcDAjAMBgNVHRMBAf8EAjAAMAoGCCqGSM49BAMD
  A2gAMGUCMQDUKGPCwh5i23dOYk6vnqlCVGmERDjTzpP/qzPXfsLZ/MbzgCR69I4y
  AoX/DFVt6HICMCdBm0NNduRDi31g7jr7tV6rsG3eh01FQCAqNK6t9AtAd6L0pxRV
  5JkNz3VtwBOLyQ==
  -----END CERTIFICATE-----
fingerprint: 5d43a3a5d7d423a90eab66b0f1363a2931ce194dfd90109cf7bb1378deb6d0a3
$ lxc list tls: --all-projects
Error: Certificate is restricted
$ lxc list tls: --project default
+------+---------+---------------------+-----------------------------------------------+-----------+-----------+
| NAME |  STATE  |        IPV4         |                     IPV6                      |   TYPE    | SNAPSHOTS |
+------+---------+---------------------+-----------------------------------------------+-----------+-----------+
| c2   | RUNNING | 10.101.45.93 (eth0) | fd42:ad83:91cc:480f:216:3eff:fe6d:2b89 (eth0) | CONTAINER | 2         |
+------+---------+---------------------+-----------------------------------------------+-----------+-----------+
$ lxc storage volume list tls:default --all-projects
Error: Certificate is restricted
$ lxc storage volume list tls:default --project default
+----------------------+------------------------------------------------------------------+-------------+--------------+---------+
|         TYPE         |                               NAME                               | DESCRIPTION | CONTENT-TYPE | USED BY |
+----------------------+------------------------------------------------------------------+-------------+--------------+---------+
| container            | c2                                                               |             | filesystem   | 1       |
+----------------------+------------------------------------------------------------------+-------------+--------------+---------+
| container (snapshot) | c2/snap0                                                         |             | filesystem   | 0       |
+----------------------+------------------------------------------------------------------+-------------+--------------+---------+
| container (snapshot) | c2/snap1                                                         |             | filesystem   | 0       |
+----------------------+------------------------------------------------------------------+-------------+--------------+---------+
| custom               | bacon                                                            |             | filesystem   | 0       |
+----------------------+------------------------------------------------------------------+-------------+--------------+---------+
| custom               | eggs                                                             |             | filesystem   | 0       |
+----------------------+------------------------------------------------------------------+-------------+--------------+---------+
| custom (snapshot)    | eggs/snap0                                                       |             | filesystem   | 0       |
+----------------------+------------------------------------------------------------------+-------------+--------------+---------+
| custom (snapshot)    | eggs/snap1                                                       |             | filesystem   | 0       |
+----------------------+------------------------------------------------------------------+-------------+--------------+---------+
| image                | e7a5ba34b253aa2b367b6c61975b3eee0aa4240f9baf0a4a4bef7bccde2637c5 |             | filesystem   | 1       |
+----------------------+------------------------------------------------------------------+-------------+--------------+---------+
```

**Comment by tomponline** [comment]
Yeah thats right @markylaing although it should probably be a bad request rather than internal server error.

**Comment by edlerd** [comment]
Not sure if this is a regression, it might not be. The motivation for filing this was as follows:

In the UI, we are using the all-projects param in two places:

1. to fetch a global list of operations
2. to connect to the events websocket for any status update on an asnc operation tirggered in the ui

In both cases, it would be nice if the all-projects param spanned to the projects that the restricted user has access to.

**Comment by tomponline** [comment]
> Assuming [76cfc85](https://github.com/canonical/lxd/commit/76cfc851a46df5958d76f1028a4247c547d26879) has now been backported to 5.21, this should be fixed to return a 401 if a restricted client uses the `all-projects` query parameter.

It should be in 5.21/edge by now.

**Comment by tomponline** [comment]
> In both cases, it would be nice if the all-projects param spanned to the projects that the restricted user has access to.

@edlerd Probably worth creating a jira request to @mionaalex so she can evaluate and prioritize it with the other feature requests.

**Comment by tomponline** [comment]
More generally what you're requesting is the same as 

https://github.com/canonical/lxd/issues/13962

**Comment by tomponline** [comment]
@edlerd shall we close this one in preference to https://github.com/canonical/lxd/issues/13962 ?

**Comment by edlerd** [comment]
> [@edlerd](https://github.com/edlerd) shall we close this one in preference to [#13962](https://github.com/canonical/lxd/issues/13962) ?

I am not sure those two are the same. I think this one is about the operation API, the other one seems like only about the events API.

**Comment by tomponline** [comment]
Noted thanks.

@markylaing do you think this issue has a different underlying cause than https://github.com/canonical/lxd/issues/13962 ?

**Comment by markylaing** [comment]
Yes. The `all-projects` query parameter for restricted TLS clients is being blocked by this function https://github.com/canonical/lxd/blob/5169fd87039da0cef8af5c054f963cfaa56df045/lxd/daemon.go#L308-L356

which was specifically added to maintain API behaviour for restricted TLS clients. I'm not sure we'll be able to change this without breaking the API.

**Comment by markylaing** [comment]
For the events API, we're also using the aforementioned function to block "all-projects" requests for restricted TLS clients. But I think the fix should be different as we'll need to evaluate the permissions on a per-event basis.

**Comment by tomponline** [comment]
> Yes. The `all-projects` query parameter for restricted TLS clients is being blocked by this function
> 
> [lxd/lxd/daemon.go](https://github.com/canonical/lxd/blob/5169fd87039da0cef8af5c054f963cfaa56df045/lxd/daemon.go#L308-L356)
> 
> Lines 308 to 356 in [5169fd8](/canonical/lxd/commit/5169fd87039da0cef8af5c054f963cfaa56df045)
> 
>  // allowProjectResourceList should be used instead of allowAuthenticated when listing resources within a project. 
>  // This prevents a restricted TLS client from listing resources in a project that they do not have access to. 
>  func allowProjectResourceList(d *Daemon, r *http.Request) response.Response { 
>  	// The caller must be authenticated. 
>  	if !auth.IsTrusted(r.Context()) { 
>  		return response.Forbidden(nil) 
>  	} 
>   
>  	isServerAdmin, err := auth.IsServerAdmin(r.Context(), d.identityCache) 
>  	if err != nil { 
>  		return response.InternalError(fmt.Errorf("Failed to determine caller privilege: %w", err)) 
>  	} 
>   
>  	// A root user can list resources in any project. 
>  	if isServerAdmin { 
>  		return response.EmptySyncResponse 
>  	} 
>   
>  	id, err := auth.GetIdentityFromCtx(r.Context(), d.identityCache) 
>  	if err != nil { 
>  		return response.InternalError(fmt.Errorf("Failed to determine caller identity: %w", err)) 
>  	} 
>   
>  	switch id.IdentityType { 
>  	case api.IdentityTypeOIDCClient: 
>  		// OIDC authenticated clients are governed by fine-grained auth. They can call the endpoint but may see an empty list. 
>  		return response.EmptySyncResponse 
>  	case api.IdentityTypeCertificateClient: 
>  		// Fine-grained TLS identities can list resources in any project. They may see an empty list. 
>  		return response.EmptySyncResponse 
>  	case api.IdentityTypeCertificateClientRestricted: 
>  		// A restricted client may be able to call the endpoint, continue. 
>  	default: 
>  		// No other identity types may list resources (e.g. metrics certificates). 
>  		return response.Forbidden(nil) 
>  	} 
>   
>  	// all-projects requests are not allowed 
>  	if shared.IsTrue(request.QueryParam(r, "all-projects")) { 
>  		return response.Forbidden(fmt.Errorf("Certificate is restricted")) 
>  	} 
>   
>  	// Disallow listing resources in projects the caller does not have access to. 
>  	if !shared.ValueInSlice(request.ProjectParam(r), id.Projects) { 
>  		return response.Forbidden(fmt.Errorf("Certificate is restricted")) 
>  	} 
>   
>  	return response.EmptySyncResponse 
>  } 
> which was specifically added to maintain API behaviour for restricted TLS clients. I'm not sure we'll be able to change this without breaking the API.
@markylaing thanks, at the very least the HTTP response code is incorrect:

> Failed to get operation permission checker: Certificate is restricted error_code: 500

Its not an internal server error, its a bad request for a TLS restricted client to use all-projects.

**Comment by tomponline** [comment]
> which was specifically added to maintain API behaviour for restricted TLS clients. I'm not sure we'll be able to change this without breaking the API.

If that was always being rejected for restricted TLS clients then thats a fair point. That being said if its always been rejected it means nobody can be using it, so we could at least potentially explore whether all-projects for restricted users means "just the projects I have access to"?

**Comment by tomponline** [comment]
> > [@markylaing](https://github.com/markylaing) thanks, at the very least the HTTP response code is incorrect:
> > > Failed to get operation permission checker: Certificate is restricted error_code: 500
> > 
> > 
> > Its not an internal server error, its a bad request for a TLS restricted client to use all-projects.
> 
> This is fixed no though no? The function returns a 401 Forbidden. I'll double check though.

Yes please.

If so then I think we can close this issue as something only supported by the fine grained authed users.

**Comment by tomponline** [comment]
So this is a wont fix @markylaing ?

**Comment by tomponline** [comment]
Or do we need to keep it open for fine grained users and update the title?

**Comment by tomponline** [comment]
@markylaing i think we can leave the restricted user behaviour as-is, but i wanted to know if we should leave open and clarify this issue is about listing operations for fine grained users? Or is this not an issue for fine grained users?

**Comment by markylaing** [comment]
It's not an issue for fine-grained users. They can view operations in all projects where they have the `can_view_operations` entitlement. I'll double check this now

**Comment by markylaing** [comment]
All working fine for fine-grained identities.

```
mark@RUBIX:~$ lxc auth group create my-group
mark@RUBIX:~$ lxc auth group permission add my-group project default can_view_operations
mark@RUBIX:~$ lxc auth identity create tls/mark --group my-group
<token>
mark@RUBIX:~$ lxc remote add localhost <token>
mark@RUBIX:~$ lxc config trust add --name dummy-operation
mark@RUBIX:~$ curl -k --cert ~/snap/lxd/common/config/client.crt --key ~/snap/lxd/common/config/client.key https://127.0.0.1:8443/1.0/operations?all-projects=true
{"type":"sync","status":"Success","status_code":200,"operation":"","error_code":0,"error":"","metadata":{"running":["/1.0/operations/af87e5f4-10f4-414c-917c-7ab4b882b450"]}}
```

So I think this is a wont fix. At this point I think we can recommend that people use fine-grained identities instead.
